utils/mention/src/parse/impl.rs

```diff
@@ -0,0 +1,302 @@
+    const PARTS: usize = 1;
+
+    /// The leading sigil of the mention after the leading arrow (`<`).
+    ///
+    /// In a channel mention, the sigil is `#`.
+    const SIGIL: &'static str;
+
+    /// Parse a mention out of a buffer.
+    ///
+    /// This will not search the buffer for a mention and will instead treat the
+    /// entire buffer as a mention.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use twilight_mention::ParseMention;
+    /// use twilight_model::id::{ChannelId, UserId};
+    ///
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// assert_eq!(ChannelId(123), ChannelId::parse("<#123>")?);
+    /// assert_eq!(UserId(456), UserId::parse("<@456>")?);
+    /// assert!(ChannelId::parse("not a mention").is_err());
+    /// # Ok(()) }
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// Returns [`ParseMentionError::LeadingArrow`] if the leading arrow is not
+    /// present.
+    ///
+    /// Returns [`ParseMentionError::Sigil`] if the mention type's sigil is not
+    /// present after the leading arrow.
+    ///
+    /// Returns [`ParseMentionError::TrailingArrow`] if the trailing arrow is
+    /// not present after the ID.
+    ///
+    /// [`ParseMentionError::LeadingArrow`]: enum.ParseMentionError.html#variant.LeadingArrow
+    /// [`ParseMentionError::Sigil`]: enum.ParseMentionError.html#variant.Sigil
+    /// [`ParseMentionError::TrailingArrow`]: enum.ParseMentionError.html#variant.TrailingArrow
+    fn parse(buf: &str) -> Result<Self, ParseMentionError<'_>>
+    where
+        Self: Sized;
+
+    /// Search a buffer for mentions and parse out any that are encountered.
+    ///
+    /// Unlike [`parse`], this will not error if anything that is indicative of
+    /// a mention is encountered but did not successfully parse, such as a `<`
+    /// but with no trailing mention sigil.
+    ///
+    /// [`parse`]: #tymethod.parse
+    #[must_use = "you must use the iterator to lazily parse mentions"]
+    fn iter(buf: &str) -> MentionIter<'_, Self>
+    where
+        Self: Sized,
+    {
+        MentionIter::new(buf)
+    }
+}
+
+impl ParseMention for ChannelId {
+    const SIGIL: &'static str = "#";
+
+    fn parse(buf: &str) -> Result<Self, ParseMentionError<'_>>
+    where
+        Self: Sized,
+    {
+        parse_id(buf, Self::SIGIL, Self::PARTS).map(ChannelId)
+    }
+}
+
+impl ParseMention for EmojiId {
+    const PARTS: usize = 2;
+    const SIGIL: &'static str = ":";
+
+    fn parse(buf: &str) -> Result<Self, ParseMentionError<'_>>
+    where
+        Self: Sized,
+    {
+        parse_id(buf, Self::SIGIL, Self::PARTS).map(EmojiId)
+    }
+}
+
+impl ParseMention for RoleId {
+    const SIGIL: &'static str = "@&";
+
+    fn parse(buf: &str) -> Result<Self, ParseMentionError<'_>>
+    where
+        Self: Sized,
+    {
+        parse_id(buf, Self::SIGIL, Self::PARTS).map(RoleId)
+    }
+}
+
+impl ParseMention for UserId {
+    const SIGIL: &'static str = "@";
```

Comment:
A user mention can also be `@!` I tihnk, not really sure for when it happens though.

Reply:
it can be, originally it was as indication that the user has a nickname, in reality the desktop client always uses this form regardless if the user has a nickname or not

Reply:
Done